furms-web-ui/src/main/java/io/imunity/furms/ui/VaadinBroadcaster.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2020 Bixbit s.c. All rights reserved.
+ * See LICENSE file for licensing information.
+ */
+
+package io.imunity.furms.ui;
+
+import com.vaadin.flow.shared.Registration;
+import io.imunity.furms.domain.communities.CommunityEvent;
+import io.imunity.furms.domain.projects.ProjectEvent;
+import io.imunity.furms.domain.sites.SiteEvent;
+import io.imunity.furms.domain.users.UserEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.event.EventListener;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Component;
+
+import java.lang.invoke.MethodHandles;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+@Component
+public class VaadinBroadcaster {
+	private final static Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+	private final List<Runnable> listeners = Collections.synchronizedList(new LinkedList<>());
+
+	public Registration register(Runnable listener) {
+		listeners.add(listener);
+		return () -> listeners.remove(listener);
+	}
+
+	private void broadcast() {
+		synchronized (listeners) {
+			for (Runnable listener : listeners) {
+				try {
+					listener.run();
+					LOG.info("Run Runnable");
+				} catch (Exception e) {
+					LOG.error("Runnable failed", e);
+				}
+			}
+		}
+	}
+
```

Comment:
For the above code:
Collections.synchronizedList creates a synchronized list - true. But it is implemented (just checked) not by synchronizing on the returned object (listeners) but on an internal 'mutex' object. That I was unsure on the call, but I was afraid that it will be the case - and indeed is.

This means that operations on the listeners object (add, remove) have mutually exclusive invocations: add can't run when another thread calls remove or add, etc. That alone is fine and we want it, as otherwise concurrent add and remove could lead to big troubles. 
But operations on the listeners object are not exclusive with the operations which are *synchronized* on the listeners object. So one thread can invoke listeners.add, and other thread can execute code from the block 37-44. And this will break.

One general hint: neither synchronized lists nor CoW lists are widely applicable.

What to do:
1. listeners should be a plain ArrayList (much better performance wide than the LinkedList), not synchronized - just plain.
2. both register() and broadcast() methods should be synchronized
3. (at this point: synchronization is perfect, the only problem is that complete process of broadcasting is blocking the class completely, and handling events may take noticeable time - professionally speaking the critical section is too wide) 
4. Refactor broadcast: remove synchronized keyword; split it into two methods: getListenersToBeNotified() and notifyListeners(list); make only the first as synchronized. (i.e. at the end broadcast should call two methods internally, just two lines)